container_images/pytest/main.py

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python3
+# Copyright 2020 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Execute pytest within a virtual environment."""
+
+import configparser
+import os
+import re
+import shutil
+import sys
+import tempfile
+import typing
+
+import toml
+import tox
+
+# Dependencies to install in addition to user's requested dependencies.
+_common_test_deps = ["pytest==6.0.1"]
+
+_per_interpreter_command = [
+  # Don't invoke the `pytest` binary directly, as it circumvents
+  # the virtual environment.
+  "python",
+  "-m",
+  "pytest",
+  # Write test result to stdout out, except if test passes.
+  "-ra",
+  # Ensure junit report uses xunit1, which is the syntax used
+  # by spyglass.
+  # https://github.com/GoogleCloudPlatform/testgrid/blob/master/metadata/junit/junit.go
+  "--override-ini=junit_family=xunit1",
+  # Within the junit report, add a namespace for the environment,
+  # to ensure that tests from different interpreters aren't
+  # considered duplicates.
+  "--junit-prefix={envname}",
+  # Write a new junit xml file for each interpreter. The filename pattern
+  # is defined in our Prow instance's configuration:
+  #  https://github.com/GoogleCloudPlatform/oss-test-infra/blob/cc1f0cf1ffecc0e3b75664b0d264abc6165276d1/prow/oss/config.yaml#L98
+  "--junit-xml={env:ARTIFACTS}/junit-{envname}.xml",
+]
+
+
+def setup_execution_root(pyproject_toml_file: str) -> str:
+  """Create execution directory and copy project's code to it.
+
+  Tox creates in-directory files, and its execution fails when the
+  source tree is mounted read-only.
+
+  Args:
+    pyproject_toml_file: Absolute path to the project config file. The
+      containing directory will be recursively copied to the execution root.
+
+  Returns:
+    Absolute path to execution root.
+  """
+  assert (os.path.isabs(pyproject_toml_file) and
```

Comment:
it's not recommended to use 'assert' outside of tests. i think you should use `if` and `raise` instead, with a useful message

Reply:
Here's the pattern that I was following:
- `assert` is for self-documenting pre-conditions of functions. If it fails, it means there's a programming bug. In terms of what the user sees, Python will raise an AssertionError that shows a stacktrace, eg:

```
/usr/bin/python3.7 /home/ericedens/projects/guest-test-infra/container_images/pytest/main.py
Traceback (most recent call last):
  File "/home/ericedens/projects/guest-test-infra/container_images/pytest/main.py", line 170, in <module>
    main()
  File "/home/ericedens/projects/guest-test-infra/container_images/pytest/main.py", line 152, in main
    assert os.path.isdir('/baz')
AssertionError
```

- `sys.exit("msg")` is for user-facing messages where a stack trace isn't necessary, and would create poor signal-to-noise. In a larger program, I'd probably raise an error, but catch it later and re-format the output prior to exiting the program. So, it was just a pragmatic choice given how small this is.

Given those explanations, let me know what you'd like in this codebase and I'll follow that :)

Reply:
you can't unit test methods with sys.exit in them, is the main thing. will leave to your judgement on that, if you think 'exiting the program' is an OK output of a function
i think we should replace any asserts with:
```python
if not <condition>:
  raise Exception("useful message")
```

Reply:
- Switched away from sys.exit in favor of raising an error
- In Python, `assert` is a syntactic shorthand for the snippet you have above (using `AssertionError`). To help with debugging, I added a string message